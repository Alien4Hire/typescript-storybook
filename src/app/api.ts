import {
  BaseQueryFn,
  FetchArgs,
  FetchBaseQueryError,
} from '@reduxjs/toolkit/query';

import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { logout, getCaTokenCookie } from './reducers/login/loginSlice';
import { setToastNotification } from './reducers/appSlice/appSlice';

export const baseUrl = process.env.REACT_APP_API || '/api/v1';

declare module '@reduxjs/toolkit/query' {
  interface FetchBaseQueryError {
    status: number;
    data:
    | {
      id?: string;
      location?: string;
      message: string;
      path?: string;
      status?: number;
      type?: string;
    }[]
    | {
      id?: string;
      location?: string;
      message: string;
      path?: string;
      status?: number;
      type?: string;
    };
  }
}

const baseQuery = fetchBaseQuery({
  baseUrl,
  prepareHeaders: (headers, { getState }) => {
    const token = getCaTokenCookie();
    if (token) {
      headers.set('authorization', `Bearer ${token}`);
    }
    headers.set('Content-Type', 'application/json;charset=UTF-8');
    return headers;
  },
});

const baseQueryWithErrorHandling: BaseQueryFn<
  string | FetchArgs,
  unknown,
  FetchBaseQueryError
> = async (args, api, extraOptions) => {
  let result = await baseQuery(args, api, extraOptions);
  if (result.error) {
    if (result.error.status === 401) {
      api.dispatch(logout());
      /**
       * Instead of logging out when the token is invalid, it is possible to refresh the token here.
       */
      // try to get a new token
      // const refreshResult = await baseQuery('/refreshToken', api, extraOptions);
      // if (refreshResult.data) {
      //   // store the new token
      //   api.dispatch(setNewToken(refreshResult.data));
      //   // retry the initial query
      //   result = await baseQuery(args, api, extraOptions);
      // } else {
      //   api.dispatch(logout());
      // }
    } else {
      const message = Array.isArray(result.error?.data)
        ? result.error?.data[0]?.message
        : result.error?.data?.message;
      api.dispatch(
        setToastNotification({
          type: 'error',
          open: true,
          message,
        })
      );
    }
  }
  return result;
};

export const apiBase = createApi({
  reducerPath: 'api',
  baseQuery: baseQueryWithErrorHandling,
  //List of tag associated with our data endpoints
  tagTypes: [
    'Templates',
    'ActivitySettings',
    'TemplateViewTypes',
    'ProductModels',
    'LearningObjectives',
    'Taxonomies',
    'Activities',
    'Lessons',
    'Content',
  ],
  refetchOnFocus: true,
  refetchOnReconnect: true,
  refetchOnMountOrArgChange: true,
  endpoints: () => ({}),
});
