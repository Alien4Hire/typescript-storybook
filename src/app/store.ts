import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { combineReducers } from '@reduxjs/toolkit';
import configReducer from './reducers/config/configSlice';
import loginReducer from './reducers/login/loginSlice';
import activitySettingsReducer from './reducers/activitySettings/activitySettingsSlice';
import lessonsReducer from './reducers/lessons/lessonsSlice';
import searchReducer from './reducers/search/searchSlice';
import resourceFilterReducer from './reducers/resourcesFilter/resourcesFilterSlice';
import templateReducer from './reducers/templates/templateSlice';
import appReducer from './reducers/appSlice/appSlice';
import templateViewTypesReducer from './reducers/templateViewTypes/templateViewTypesSlice';
import targetFolderSideBarReducer from './reducers/targetFolderSideBar/targetFolderSideBarSlice';

import { setupListeners } from '@reduxjs/toolkit/query/react';
import { apiBase } from './api';

const rootReducer = combineReducers({
  config: configReducer,
  login: loginReducer,
  activitySettings: activitySettingsReducer,
  lessons: lessonsReducer,
  templates: templateReducer,
  search: searchReducer,
  resourcesFilter: resourceFilterReducer,
  app: appReducer,
  templateViewTypes: templateViewTypesReducer,
  targetFolderSideBar: targetFolderSideBarReducer,
  [apiBase.reducerPath]: apiBase.reducer,
});

// Use also for store setup in jest unit tests
// Use-case: unit test needs preloadedState
// Example: client/src/components/Pages/LoginPage/LoginPage.test.tsx
export const setupStore = (preloadedState: object = {}) => {
  return configureStore({
    reducer: rootReducer,
    devTools: process.env.NODE_ENV !== 'production',
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(apiBase.middleware),
    preloadedState,
  });
};

export const store = setupStore();

setupListeners(store.dispatch);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
