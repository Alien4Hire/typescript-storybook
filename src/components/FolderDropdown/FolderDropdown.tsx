import {
  DropdownMenu,
  MlDropdownMenuOptionPropsT,
} from '../DropdownMenu/DropdownMenu';
import { foldersForDropdown, FolderT } from '../../util/folderTree';
import { useAppDispatch, useAppSelector } from '../../app/hooks';
import {
  getInitialLessons,
  getLessons,
} from '../../app/reducers/lessons/lessonsSlice';
import { useEffect } from 'react';

export type FolderDropdownPropsT = {
  open: boolean;
  anchorEl: HTMLElement | null;
  onItemClick: (folder: MlDropdownMenuOptionPropsT) => void;
  onClose: () => void;
};

const FolderDropdown = ({
  open,
  anchorEl,
  onItemClick,
  onClose,
}: FolderDropdownPropsT) => {
  const dispatch = useAppDispatch();

  // Temp until wired up with RTKquery to backend
  useEffect(() => {
    dispatch(getInitialLessons());
  }, [dispatch]);

  const { items: folders } = useAppSelector(getLessons);
  const options = foldersForDropdown(folders as FolderT);

  const handleItemClick = (selectedValue: string) => {
    const folder = options.find((o) => o.value === selectedValue);
    if (folder) {
      onItemClick(folder as MlDropdownMenuOptionPropsT);
    }
  };

  return (
    <DropdownMenu
      open={open}
      anchorEl={anchorEl}
      options={options}
      onItemClick={handleItemClick}
      onClose={onClose}
    />
  );
};

export default FolderDropdown;
