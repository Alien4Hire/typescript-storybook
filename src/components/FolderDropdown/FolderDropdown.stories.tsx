import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import FolderDropdown from './FolderDropdown';
import { store } from '../../app/store';
import { IconButton } from '@mui/material';
import AddIcon from '@mui/icons-material/Add';

export default {
  title: 'Component/FolderDropdown',
  component: FolderDropdown,
} as ComponentMeta<typeof FolderDropdown>;

const Template: ComponentStory<typeof FolderDropdown> = () => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);

  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleItemClick = () => {};

  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <Provider store={store}>
      <Router>
        <IconButton aria-label="more" aria-haspopup onClick={handleClick}>
          <AddIcon />
        </IconButton>
        <FolderDropdown
          open={Boolean(anchorEl)}
          anchorEl={anchorEl}
          onItemClick={handleItemClick}
          onClose={handleClose}
        />
      </Router>
    </Provider>
  );
};

export const Primary = Template.bind({});
Primary.args = {};
