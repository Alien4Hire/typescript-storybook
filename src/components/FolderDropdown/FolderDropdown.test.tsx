import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import FolderDropdown from './FolderDropdown';
import { store } from '../../app/store';
import '../../setupTests';

const baseProps = {
  open: true,
  anchorEl: document.createElement('div'),
  onItemClick: jest.fn(),
  onClose: jest.fn(),
};

const renderComponent = () => {
  return render(
    <Provider store={store}>
      <Router>
        <FolderDropdown {...baseProps} />
      </Router>
    </Provider>
  );
};

it('renders without crashing', () => {
  const view = renderComponent();
  expect(view).toMatchSnapshot();
});
