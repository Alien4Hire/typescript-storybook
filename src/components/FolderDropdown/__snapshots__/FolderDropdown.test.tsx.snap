// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c5 {
  background-color: #fff;
  color: rgba(0,0,0,0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  border-radius: 4px;
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
}

.c3 {
  background-color: #fff;
  color: rgba(0,0,0,0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
}

.c14 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

.c9 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
}

.c9::-moz-focus-inner {
  border-style: none;
}

.c9.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

.c13 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  display: block;
}

.c1 {
  position: fixed;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
  background-color: transparent;
  -webkit-tap-highlight-color: transparent;
}

.c0 {
  position: fixed;
  z-index: 1300;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
}

.c2 {
  z-index: -1;
}

.c7 {
  list-style: none;
  margin: 0;
  padding: 0;
  position: relative;
  padding-top: 8px;
  padding-bottom: 8px;
}

.c11 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  min-width: 0;
  margin-top: 4px;
  margin-bottom: 4px;
}

.c6 {
  position: absolute;
  overflow-y: auto;
  overflow-x: hidden;
  min-width: 16px;
  min-height: 16px;
  max-width: calc(100% - 32px);
  max-height: calc(100% - 32px);
  outline: 0;
}

.c4 {
  max-height: calc(100% - 96px);
  -webkit-overflow-scrolling: touch;
}

.c8 {
  outline: 0;
}

.c10 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 48px;
  padding-top: 6px;
  padding-bottom: 6px;
  box-sizing: border-box;
  white-space: nowrap;
  padding-left: 16px;
  padding-right: 16px;
}

.c10:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(0,0,0,0.04);
}

.c10.Mui-selected {
  background-color: rgba(25,118,210,0.08);
}

.c10.Mui-selected.Mui-focusVisible {
  background-color: rgba(25,118,210,0.2);
}

.c10.Mui-selected:hover {
  background-color: rgba(25,118,210,0.12);
}

.c10.Mui-focusVisible {
  background-color: rgba(0,0,0,0.12);
}

.c10.Mui-disabled {
  opacity: 0.38;
}

.c10 + .MuiDivider-root {
  margin-top: 8px;
  margin-bottom: 8px;
}

.c10 + .MuiDivider-inset {
  margin-left: 52px;
}

.c10 .MuiListItemText-root {
  margin-top: 0;
  margin-bottom: 0;
}

.c10 .MuiListItemText-inset {
  padding-left: 36px;
}

.c10 .MuiListItemIcon-root {
  min-width: 36px;
}

.c12 {
  color: #080808 !important;
  margin-left: 36px;
  padding-left: 0;
}

@media print {
  .c9 {
    color-adjust: exact;
  }
}

@media (hover:none) {
  .c10:hover {
    background-color: transparent;
  }
}

@media (hover:none) {
  .c10.Mui-selected:hover {
    background-color: rgba(25,118,210,0.08);
  }
}

@media (min-width:600px) {
  .c10 {
    min-height: auto;
  }
}

<body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="c0 MuiModal-root MuiPopover-root-kLYbPy MuiPopover-root MuiMenu-root-egkglY MuiMenu-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="c1 MuiBackdrop-root MuiBackdrop-invisible c2"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="c3 MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 c4 MuiPaper-root-jrxuSB c5 MuiMenu-paper MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 c6 MuiPopover-paper"
        style="opacity: 1; transform: scale(1, 1); max-height: 216px; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
        tabindex="-1"
      >
        <ul
          aria-labelledby="long-button"
          class="c7 MuiList-root MuiList-padding c8 MuiMenu-list"
          role="menu"
          tabindex="-1"
        >
          <li
            class="c9 MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root Mui-focusVisible c10"
            role="menuitem"
            tabindex="0"
          >
            <div
              class="c11 MuiListItemText-root c12"
              color="#080808"
              option="[object Object]"
            >
              <span
                class="c13 MuiTypography-root MuiTypography-body1 MuiListItemText-primary"
              >
                Root
              </span>
            </div>
            <span
              class="c14 MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": @media print {

}

@media (hover:none) {

}

@media (hover:none) {

}

@media (min-width:600px) {

}

<div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
